EclipseLink IllegalArgumentException on JPQL update of embeddable attribute that has another embeddable

The test has an Entity with a hierarchy of Embeddables (Entity -> Embeddable -> Embeddable -> Embeddable) and performs a JPQL update statement UPDATE Business b SET b.location=?1, b.name=?2 WHERE b.id=?3 to modify the highest level Embeddable attribute under the entity and one other entity attribute. In processing this JPQL statement, EclipseLink appears to get confused when trying to obtain the value of a field of the depth 2 Embeddable, attempting to invoke it on a target of the depth 1 Embeddable instead. The failure is:

```
jakarta.persistence.PersistenceException: Exception [EclipseLink-26] (Eclipse Persistence Services - 5.0.0-B08.v202505280949-dfc411d38767f696ce9f2741de051aef050cbeba): org.eclipse.persistence.exceptions.DescriptorException
Exception Description: Trying to get value for instance variable [street] of type [test.jakarta.data.jpa.web.Street] from the object [test.jakarta.data.jpa.web.Location]. The specified object is not an instance of the class or interface declaring the underlying field.
Internal Exception: java.lang.IllegalArgumentException: Can not get test.jakarta.data.jpa.web.Street field test.jakarta.data.jpa.web.Address.street on test.jakarta.data.jpa.web.Location
Mapping: org.eclipse.persistence.mappings.AggregateObjectMapping[street]
Descriptor: RelationalDescriptor(test.jakarta.data.jpa.web.Address --> [DatabaseTable(Business)])
at org.eclipse.persistence.internal.jpa.QueryImpl.executeUpdate(QueryImpl.java:310)
at io.openliberty.data.internal.persistence.QueryInfo.execute(QueryInfo.java:1696)
at io.openliberty.data.internal.persistence.RepositoryImpl.invoke(RepositoryImpl.java:572)
Caused by: Exception [EclipseLink-26] (Eclipse Persistence Services - 5.0.0-B08.v202505280949-dfc411d38767f696ce9f2741de051aef050cbeba): org.eclipse.persistence.exceptions.DescriptorException
Exception Description: Trying to get value for instance variable [street] of type [test.jakarta.data.jpa.web.Street] from the object [test.jakarta.data.jpa.web.Location]. The specified object is not an instance of the class or interface declaring the underlying field.
Internal Exception: java.lang.IllegalArgumentException: Can not get test.jakarta.data.jpa.web.Street field test.jakarta.data.jpa.web.Address.street on test.jakarta.data.jpa.web.Location
Mapping: org.eclipse.persistence.mappings.AggregateObjectMapping[street]
Descriptor: RelationalDescriptor(test.jakarta.data.jpa.web.Address --> [DatabaseTable(Business)])
at org.eclipse.persistence.exceptions.DescriptorException.illegalArgumentWhileGettingValueThruInstanceVariableAccessor(DescriptorException.java:653)
at org.eclipse.persistence.internal.descriptors.InstanceVariableAttributeAccessor.lambda$getAttributeValueFromObject$1(InstanceVariableAttributeAccessor.java:73)
at org.eclipse.persistence.internal.security.PrivilegedAccessHelper.callDoPrivilegedWithException(PrivilegedAccessHelper.java:260)
at org.eclipse.persistence.internal.descriptors.InstanceVariableAttributeAccessor.getAttributeValueFromObject(InstanceVariableAttributeAccessor.java:69)
at org.eclipse.persistence.mappings.DatabaseMapping.getAttributeValueFromObject(DatabaseMapping.java:692)
at org.eclipse.persistence.internal.expressions.ParameterExpression.getValue(ParameterExpression.java:273)
at org.eclipse.persistence.internal.databaseaccess.DatabaseCall.translate(DatabaseCall.java:1139)
at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:281)
at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:268)
at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeNoSelectCall(DatasourceCallQueryMechanism.java:325)
at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.updateAll(DatasourceCallQueryMechanism.java:940)
at org.eclipse.persistence.queries.UpdateAllQuery.executeDatabaseQuery(UpdateAllQuery.java:159)
at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:934)
at org.eclipse.persistence.queries.DatabaseQuery.executeInUnitOfWork(DatabaseQuery.java:833)
at org.eclipse.persistence.queries.ModifyAllQuery.executeInUnitOfWork(ModifyAllQuery.java:153)
at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3018)
at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1866)
at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1848)
at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1813)
at org.eclipse.persistence.internal.jpa.QueryImpl.executeUpdate(QueryImpl.java:303)
Caused by: java.lang.IllegalArgumentException: Can not get test.jakarta.data.jpa.web.Street field test.jakarta.data.jpa.web.Address.street on test.jakarta.data.jpa.web.Location
at java.base/jdk.internal.reflect.MethodHandleFieldAccessorImpl.newGetIllegalArgumentException(MethodHandleFieldAccessorImpl.java:86)
at java.base/jdk.internal.reflect.MethodHandleObjectFieldAccessorImpl.get(MethodHandleObjectFieldAccessorImpl.java:61)
at java.base/java.lang.reflect.Field.get(Field.java:444)
at org.eclipse.persistence.internal.descriptors.InstanceVariableAttributeAccessor.lambda$getAttributeValueFromObject$0(InstanceVariableAttributeAccessor.java:70)
at org.eclipse.persistence.internal.security.PrivilegedAccessHelper.callDoPrivilegedWithException(PrivilegedAccessHelper.java:191)
at org.eclipse.persistence.internal.security.PrivilegedAccessHelper.callDoPrivilegedWithException(PrivilegedAccessHelper.java:258)
```
Here are the entities and embeddables involved,

Entity:

```
@Entity
public class Business {

    public String name;

    @GeneratedValue
    @Id
    public int id;

    @Embedded
    public Location location;

    public Business() {
    }

    Business(float latitude, float longitude, String city, String state, int zip,
             int houseNum, String streetName, String streetDir, String name) {
        Street street = new Street(streetName, streetDir);
        Address address = new Address(city, state, zip, houseNum, street);
        this.name = name;
        this.location = new Location(address, latitude, longitude);
    }
}
```
Location Embeddable:

```
@Embeddable
public class Location {

    @Embedded
    public Address address;

    @Column(columnDefinition = "DECIMAL(8,5) NOT NULL")
    public float latitude;

    @Column(columnDefinition = "DECIMAL(8,5) NOT NULL")
    public float longitude;

    public Location() {
    }

    public Location(Address address, float latitude, float longitude) {
        this.address = address;
        this.latitude = latitude;
        this.longitude = longitude;
    }
}
```
Address Embeddable:

```
@Embeddable
public class Address {

    public String city;

    public int houseNum;

    public String state;

    @Embedded
    public Street street;

    @Convert(converter = ZipCodeConverter.class)
    public ZipCode zip;

    public Address() {
    }

    Address(String city, String state, int zip, int houseNum, Street street) {
        this.city = city;
        this.state = state;
        this.zip = ZipCode.of(zip);
        this.houseNum = houseNum;
        this.street = street;
    }
}
```
Street Embeddable:

```
@Embeddable
public class Street {

    public String direction;

    @Column(name = "STREETNAME")
    public String name;

    public Street() {
    }

    public Street(String name, String direction) {
        this.name = name;
        this.direction = direction;
    }
}
```

--- a/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/ParameterExpression.java
+++ b/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/ParameterExpression.java
@@ -270,7 +270,7 @@ public class ParameterExpression extends BaseExpression {
                     if (mapping != null) {
                         //Nested attribute
                         if (value.getClass() != descriptor.getJavaClass() && this.getLocalBase().isQueryKeyExpression() && ((QueryKeyExpression)this.getLocalBase()).getBaseExpression().isQueryKeyExpression()) {
-                            value = ((QueryKeyExpression)((QueryKeyExpression)this.getLocalBase()).getBaseExpression()).getMapping().getAttributeValueFromObject(value);
+                            value = getParentNodeValue(descriptor.getJavaClass(), (QueryKeyExpression)this.getLocalBase(), value);
                         }
                         if (value != null) {
                             value = mapping.getRealAttributeValueFromObject(value, session);
@@ -330,6 +330,34 @@ public class ParameterExpression extends BaseExpression {
             value = this.localBase.getFieldValue(value, session);
         }
 
+        return value;
+    }
+
+    private boolean directionUp = true;
+
+    private Object getParentNodeValue(Class<?> parentNodeClass, QueryKeyExpression queryKeyExpression, Object inputValue) {
+        Object value = null;
+        QueryKeyExpression baseExpression = queryKeyExpression;
+        DatabaseMapping mapping = baseExpression.getMapping();
+        if (directionUp) {
+            baseExpression = (QueryKeyExpression) queryKeyExpression.getBaseExpression();
+            mapping = baseExpression.getMapping();
+        }
+        if (mapping.getReferenceDescriptor().getJavaClass() == parentNodeClass &&
+                mapping.getDescriptor().getJavaClass() == inputValue.getClass()) {
+            value = mapping.getAttributeValueFromObject(inputValue);
+        } else {
+            Object tmpValue = inputValue;
+            if (mapping.getDescriptor().getJavaClass() == inputValue.getClass()) {
+                tmpValue = mapping.getAttributeValueFromObject(inputValue);
+            }
+            if (tmpValue != inputValue) {
+                directionUp = false;
+                return tmpValue;
+            }
+            tmpValue = getParentNodeValue(parentNodeClass, baseExpression, tmpValue);
+            value = mapping.getAttributeValueFromObject(tmpValue);
+        }
         return value;
     }
 

--- a/jpa/eclipselink.jpa.test.jse/src/it/java/org/eclipse/persistence/jpa/embeddable/TestNestedEmbeddable.java
+++ b/jpa/eclipselink.jpa.test.jse/src/it/java/org/eclipse/persistence/jpa/embeddable/TestNestedEmbeddable.java
@@ -126,7 +126,7 @@ public class TestNestedEmbeddable {
     }
 
     @Test
-    public void JPQLNestedEmbeddableTests() {
+    public void JPQLNestedEmbeddableSelectTests() {
         EntityManager em = emf2.createEntityManager();
         try {
             // Test to make sure that moving around the dot notation for nested embeddables is valid and doesn't change the query
@@ -141,6 +141,63 @@ public class TestNestedEmbeddable {
             queryEmbed.getResultList();
             Assert.assertEquals(1, _sql.size());
             Assert.assertEquals("SELECT CITY FROM SPECEMPLOYEE WHERE (ZIPCODE = ?)", _sql.remove(0));
+        } finally {
+            if (em.isOpen()) {
+                em.close();
+            }
+        }
+    }
+
+    @Test
+    public void JPQLNestedEmbeddableUpdateTest() {
+        if (emf == null)
+            return;
+        EntityManager em = emf.createEntityManager();
+        try {
+            //Prepare data for a test
+            em.getTransaction().begin();
+            Order o = new Order();
+
+            String billingZip = "6789";
+            String shippingZip ="9876";
+
+            OrderPK opk = new OrderPK();
+            opk.setBillingAddress(new Address(new Zipcode(billingZip)));
+            opk.setShippingAddress(new Address(new Zipcode(shippingZip)));
+
+            o.setId(opk);
+
+            em.persist(o);
+            em.getTransaction().commit();
+            em.clear();
+
+            Order foundOrder = em.find(Order.class, o.getId());
+            Assert.assertNotNull(foundOrder);
+
+            em.getTransaction().begin();
+
+            String newBillingZip = "1111";
+            String newShippingZip ="2222";
+
+            OrderPK opkNew = new OrderPK();
+            opkNew.setBillingAddress(new Address(new Zipcode(newBillingZip)));
+            opkNew.setShippingAddress(new Address(new Zipcode(newShippingZip)));
+            //Tests nested Embedables as JPQL parameter in UPDATE query
+            int updatedCount = em.createQuery("UPDATE Order o SET o.id = :newId WHERE o.id = :id")
+                    .setParameter("newId", opkNew)
+                    .setParameter("id", foundOrder.getId())
+                    .executeUpdate();
+            Assert.assertEquals(1, updatedCount);
+            em.getTransaction().commit();
+            em.clear();
+
+            //Tests nested Embedables as JPQL parameter in SELECT query
+            Order orderResult = em.createQuery("SELECT o FROM Order o WHERE o.id = :newId", Order.class)
+                    .setParameter("newId", opkNew)
+                    .getSingleResult();
+            Assert.assertNotNull(orderResult);
+            Assert.assertEquals(opkNew, orderResult.getId());
+
         } finally {
             if (em.isOpen()) {
                 em.close();

